grug/testing/src/client.rs

```diff
@@ -0,0 +1,313 @@
+    PP: ProposalPreparer + Send + Sync,
+    ID: Indexer + Send + Sync,
+{
+    type Error = anyhow::Error;
+
+    async fn search_tx(&self, hash: Hash256) -> Result<SearchTxOutcome, Self::Error> {
+        self.txs
+            .lock()
+            .await
+            .get(&hash)
+            .cloned()
+            .ok_or(anyhow::anyhow!("tx not found: {hash}"))
+    }
+}
+
+#[async_trait]
+impl<DB, VM, PP, ID> BroadcastClient for MockClient<DB, VM, PP, ID>
+where
+    DB: Db + Send + Sync,
+    VM: Vm + Clone + Send + Sync + 'static,
+    PP: ProposalPreparer + Send + Sync,
+    ID: Indexer + Send + Sync,
+    AppError: From<DB::Error> + From<VM::Error> + From<PP::Error> + From<ID::Error>,
+{
+    type Error = anyhow::Error;
+
+    async fn broadcast_tx(&self, tx: Tx) -> Result<BroadcastTxOutcome, Self::Error> {
+        let tx_hash = tx.tx_hash()?;
+
+        let check_tx = self.suite.lock().await.app.do_check_tx(tx.clone())?;
+        if check_tx.result.is_err() {
+            return Ok(BroadcastTxOutcome { tx_hash, check_tx });
+        };
+
+        match &self.block_mode {
+            BlockModeCache::Timed { buffer, .. } => {
+                buffer.lock().await.push(tx);
+            },
+            BlockModeCache::OnBroadcast => {
+                next_block(
+                    self.txs.lock().await.deref_mut(),
+                    self.blocks.lock().await.deref_mut(),
+                    self.suite.lock().await.deref_mut(),
+                    vec![tx],
+                )
+                .await?;
+            },
+        }
+
+        Ok(BroadcastTxOutcome { tx_hash, check_tx })
+    }
+}
+
+async fn index(
+    index_txs: &mut BTreeMap<Hash256, SearchTxOutcome>,
+    index_blocks: &mut BTreeMap<u64, (Block, BlockOutcome)>,
+    outcome: MakeBlockOutcome,
+    block_info: BlockInfo,
+) -> anyhow::Result<()> {
+    for (index, ((tx, hash), outcome)) in outcome
+        .txs
+        .iter()
+        .zip(outcome.block_outcome.tx_outcomes.iter())
+        .enumerate()
+    {
+        index_txs.insert(*hash, SearchTxOutcome {
+            hash: *hash,
+            height: block_info.height,
+            index: index as u32,
+            tx: tx.clone(),
+            outcome: outcome.clone(),
+        });
+    }
+
+    let block = Block {
+        info: block_info,
+        txs: outcome.txs,
+    };
+
+    index_blocks.insert(block_info.height, (block, outcome.block_outcome));
+
+    Ok(())
+}
+
+async fn next_block<DB, VM, PP, ID>(
+    index_txs: &mut BTreeMap<Hash256, SearchTxOutcome>,
+    index_blocks: &mut BTreeMap<u64, (Block, BlockOutcome)>,
+    suite: &mut TestSuite<DB, VM, PP, ID>,
+    tx: Vec<Tx>,
+) -> anyhow::Result<()>
+where
+    DB: Db,
+    VM: Vm + Clone + 'static,
+    PP: ProposalPreparer,
+    ID: Indexer,
+    AppError: From<DB::Error> + From<VM::Error> + From<PP::Error> + From<ID::Error>,
+{
+    let outcome = suite.make_block(tx);
+    index(index_txs, index_blocks, outcome, suite.block).await
```

Comment:
There is no need for `index` to be `async` (see other comment), meaning `next_block` doesn't need to be `async` either.